Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferProxy.h

```diff
@@ -101,14 +98,28 @@ class RemoteImageBufferProxy : public WebCore::ImageBuffer {
     void assertDispatcherIsCurrent() const;
 
     DisplayListRecorderFlushIdentifier m_sentFlushIdentifier;
-    Lock m_receivedFlushIdentifierLock;
-    Condition m_receivedFlushIdentifierChangedCondition;
-    DisplayListRecorderFlushIdentifier m_receivedFlushIdentifier WTF_GUARDED_BY_LOCK(m_receivedFlushIdentifierLock); // Only modified on the main thread but may get queried on a secondary thread.
+    Ref<RemoteImageBufferProxyFlushState> m_flushState;
     WeakPtr<RemoteRenderingBackendProxy> m_remoteRenderingBackendProxy;
     RemoteDisplayListRecorderProxy m_remoteDisplayList;
     bool m_needsFlush { false };
 };
 
+class RemoteImageBufferProxyFlushState : public ThreadSafeRefCounted<RemoteImageBufferProxyFlushState> {
+    WTF_MAKE_NONCOPYABLE(RemoteImageBufferProxyFlushState);
+    WTF_MAKE_FAST_ALLOCATED;
+public:
+    RemoteImageBufferProxyFlushState() = default;
+    void waitForDidFlushOnSecondaryThread(DisplayListRecorderFlushIdentifier);
+    void markCompletedFlush(DisplayListRecorderFlushIdentifier);
+    void cancel();
+    DisplayListRecorderFlushIdentifier completedFlush() const;
```

Comment:
The function name sounds like it would return a `bool` or `void`, not an identifier. Is this `identifierForCompletedFlush()`?